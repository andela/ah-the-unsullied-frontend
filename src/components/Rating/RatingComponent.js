import React, {Component} from 'react';
import Rating from  'react-star-rating-lite';
import {connect} from 'react-redux';
import PropTypes from 'prop-types';
import {
  postRating,
  getAverageRating
} from '../../actions/Rating/ratingActions';

export class ArticleRating extends Component{

  componentDidMount() {
    const {getAverageRating} = this.props;
    const {slug} = this.props;
    getAverageRating(slug)

  }

   onClickHandler = (ratingValue) => {
    const ratingData = {
      'rating': ratingValue
    };

    const {slug} = this.props;
    const {postRating} = this.props;
    postRating(ratingData, slug);
};

  render() {
  const { rating } = this.props;
  const { auth } = this.props;
  const { article } =this.props;
  const articleAuthor = article.article.article.author.username;
  if(!auth.isAuthenticated){
    return(
      <div className="container">
        <p> Please login to rate this article</p>
      <Rating
        value={rating}
        color="orange"
        weight="25"
        readonly
      />
      </div>
    )
  }
  else if(articleAuthor === auth.user.username){
    return (
      <div className="container">
        <p> Average rating:</p>
        <Rating
          value={rating}
          color="orange"
          weight="25"
          readonly
        />
      </div>
    )
  }
  else {
    return (
      <div className="container">
      <Rating
        value={rating}
        color="orange"
        weight="25"
        onClick={this.onClickHandler}
      />
      </div>
    )
  }
}
}

ArticleRating.propTypes = {
  getAverageRating: PropTypes.func.isRequired,
  rating: PropTypes.object.isRequired,
  postRating: PropTypes.func.isRequired,
  slug: PropTypes.object.isRequired
};

export const mapStateToProps = state => {
  return{
    auth: state.auth,
    article: state.articlereducer,
    ratingError: state.rating
  }
};
export default connect(mapStateToProps, { getAverageRating, postRating })(ArticleRating);
