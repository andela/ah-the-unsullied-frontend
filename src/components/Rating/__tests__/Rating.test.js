import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import Enzyme, { mount, shallow } from 'enzyme';
import { ArticleRating, mapStateToProps } from '../RatingComponent';

Enzyme.configure({ adapter: new Adapter() });

describe('Rating component', () => {
  let props;
  let wrapper;
  let wrapperInstance;
  let state;

  beforeEach(() => {
    props = {
      slug: {
        slug: 'article-slug'
      },
      rating: {
        rating: 5
      },
      postRating: jest.fn(() => {
        Promise.resolve();
      }),
      getAverageRating: jest.fn(() => {
        Promise.resolve();
      }),
      auth: {
        isAuthenticated: true,
        user: {
          username: 'allan'
        }
      },
      article: {
        article: {
          article: {
            author: {
              username: 'nesh'
            }
          }
        }
      }
    };
    state = {
      articlereducer: {
        article: {}
      },
      auth: {
        isAuthenticated: true,
        user: {
          username: 'allan'
        }
      },
      rating: {
        error: 'a problem here'
      }
    };
    wrapper = shallow(<ArticleRating {...props} />);

    wrapperInstance = wrapper.instance();
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('onClickHandler', () => {
    wrapperInstance.onClickHandler(5);
    expect(props.postRating).toBeCalledWith(
      { rating: 5 },
      { slug: 'article-slug' }
    );
  });

  it('componentDidMount', () => {
    const spy = jest.spyOn(ArticleRating.prototype, 'componentDidMount');
    mount(<ArticleRating {...props} />);
    expect(spy).toHaveBeenCalled();
    expect(props.getAverageRating).toBeCalledWith({ slug: 'article-slug' });
  });

  it('mapStateToProps', () => {
    const propsState = mapStateToProps(state);
    expect(propsState.article).toEqual(state.articlereducer);
    expect(propsState.auth).toEqual(state.auth);
    expect(propsState.ratingError).toEqual(state.rating);
  });
});
