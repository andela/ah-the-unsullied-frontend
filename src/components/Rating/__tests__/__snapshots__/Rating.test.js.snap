// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rating component should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ArticleRating
    article={
      Object {
        "article": Object {
          "article": Object {
            "author": Object {
              "username": "nesh",
            },
          },
        },
      }
    }
    auth={
      Object {
        "isAuthenticated": true,
        "user": Object {
          "username": "allan",
        },
      }
    }
    getAverageRating={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "slug": "article-slug",
            },
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    postRating={[MockFunction]}
    rating={
      Object {
        "rating": 5,
      }
    }
    slug={
      Object {
        "slug": "article-slug",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <t
        color="orange"
        onClick={[Function]}
        value={
          Object {
            "rating": 5,
          }
        }
        weight="25"
      />,
      "className": "container",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "color": "orange",
        "onClick": [Function],
        "value": Object {
          "rating": 5,
        },
        "weight": "25",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <t
          color="orange"
          onClick={[Function]}
          value={
            Object {
              "rating": 5,
            }
          }
          weight="25"
        />,
        "className": "container",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "color": "orange",
          "onClick": [Function],
          "value": Object {
            "rating": 5,
          },
          "weight": "25",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
